crates/table/src/page_pool.rs

```diff
@@ -0,0 +1,294 @@
+    pub fn put_many(&self, pages: impl Iterator<Item = Box<Page>>) {
+        for page in pages {
+            self.put(page);
+        }
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports fixed rows of size `fixed_row_size`.
+    pub fn take_with_fixed_row_size(&self, fixed_row_size: Size) -> Box<Page> {
+        self.inner.take_with_fixed_row_size(fixed_row_size)
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports a maximum of `max_rows_in_page` rows.
+    fn take_with_max_row_count(&self, max_rows_in_page: usize) -> Box<Page> {
+        self.inner.take_with_max_row_count(max_rows_in_page)
+    }
+
+    /// Deserialize a page from `buf` but reuse the allocations in the pool.
+    pub fn take_deserialize_from(&self, buf: &[u8]) -> Result<Box<Page>, DecodeError> {
+        self.deserialize(bsatn::Deserializer::new(&mut &*buf))
+    }
+
+    /// Returns the number of pages outside the pool.
+    ///
+    /// Note that if a page is dropped outside the pool,
+    /// it won't be aware and will count that as a leak.s
+    pub fn unpooled_pages_count(&self) -> usize {
+        self.inner.unpooled_pages_count.load(Ordering::Relaxed)
+    }
+
+    /// Returns the number of pages dropped by the pool because the pool was at capacity.
+    pub fn dropped_pages_count(&self) -> usize {
+        self.inner.dropped_pages_count.load(Ordering::Relaxed)
+    }
+}
+
+impl<'de> DeserializeSeed<'de> for &PagePool {
+    type Output = Box<Page>;
+
+    fn deserialize<D: Deserializer<'de>>(self, de: D) -> Result<Self::Output, D::Error> {
+        de.deserialize_product(self)
+    }
+}
+
+impl<'de> ProductVisitor<'de> for &PagePool {
+    type Output = Box<Page>;
+
+    fn product_name(&self) -> Option<&str> {
+        Some("Page")
+    }
+
+    fn product_len(&self) -> usize {
+        2
+    }
+
+    fn visit_seq_product<A: SeqProductAccess<'de>>(self, mut prod: A) -> Result<Self::Output, A::Error> {
+        let header = prod
+            .next_element::<PageHeader>()?
+            .ok_or_else(|| A::Error::invalid_product_length(2, &self))?;
+        let row_data = prod
+            .next_element()?
+            .ok_or_else(|| A::Error::invalid_product_length(2, &self))?;
+
+        // TODO(perf, centril): reuse the allocation of `present_rows` in `page`.
+        let mut page = self.take_with_max_row_count(header.max_rows_in_page());
+        // SAFETY: `header` and `row_data` are consistent with each other.
+        unsafe { page.set_raw(header, row_data) };
+
+        Ok(page)
+    }
+
+    fn visit_named_product<A: NamedProductAccess<'de>>(self, _: A) -> Result<Self::Output, A::Error> {
+        unreachable!()
+    }
+}
+
+/// The inner actual page pool containing all the logic.
+struct PagePoolInner {
+    pages: ArrayQueue<Box<Page>>,
+    unpooled_pages_count: AtomicUsize,
+    dropped_pages_count: AtomicUsize,
```

Comment:
In addition to `dropped_pages_count`, I think tracking three other metrics would be helpful to show us how useful this is:
`new_pages_allocated` - to track how often we needed to make a fresh page, and
`pages_reused` - to track how many times `take` returned something from the pool
`pages_returned` - to track how many deallocations were avoided. I think deallocating pages was taking more time than allocating them in recent tests.

Those metrics would tell how many operations are being saved by the pool, and how many could potentially be saved if there were more space in the pool.